src/lib/gateway.ts

```diff
@@ -0,0 +1,220 @@
+import axios, { AxiosRequestConfig } from "axios";
+import { getDb, getToken } from "../store/db";
+import { getGatewayUrl } from "./urls";
+
+export const gatewayClient = axios.create({
+  headers: {
+    "Content-Type": "application/json",
+  },
+});
+
+/**
+ * Request interceptor to pass in JWT bearer token
+ *
+ */
+gatewayClient.interceptors.request.use(
+  async (config) => {
+    const token = getToken();
+    const gatewayUrl = getGatewayUrl();
+
+    config.baseURL = gatewayUrl;
+
+    if (token) {
+      config.headers = {
+        ...config.headers,
+        authorization: `Bearer ${token}`,
+      };
+    }
+
+    return config;
+  },
+  (error) => Promise.reject(error)
+);
+
+/**
+ * Response interceptor
+ *
+ */
+gatewayClient.interceptors.response.use(
+  (response) => response,
+  async (error) => {
+    if (
+      error.response.status === 401 &&
+      error.response.data &&
+      error.response.data.error === "Unauthorized"
+    ) {
+      return Promise.reject(
+        new Error(
+          "Authorization token has expired, run `bls login` to re-authenticate."
+        )
+      );
+    }
+
+    return Promise.reject(error);
+  }
+);
+
+export type GatewayEndpoints = {
+  "[GET] /functions": { params: {} };
```

Comment:
I do see repetitions of these keys/signatures across the codebase.

Would it be worthwhile to declare these are constants and refer the the constants instead?
If not, im assuming we can be very sure that the ts type-checking will be propagated throughout the codebase where the vars will be used (to prevent typos in strings).

Reply:
@zees-dev Yeah, these are GatewayEndpoint types are equivalent to referring a constant variable, when it comes to preventing misuse and typos. All usage for the endpoints is tied back to these types, and any usage that doesn't match the defined endpoint or its required parameters would not compile. Let me know if you have other questions. Thanks for reviewing this, appreciate it!